VideoBackends: Wayland support

stenzek opened the pull request:
Currently, Dolphin can semi-run the Qt frontend against the Wayland QPA, however, our video backends have no such support, and just crash out.

Most of the work to support this configuration was done as part of #7450, this PR just adds Wayland WSI integration to the backends, and a basic Wayland platform for NoGUI.

adelpozoman commented:
If you really can make this it would be awesome

stenzek commented:
Input support added, as well as a basic platform for NoGUI. Line count is up again due to being based on PR7512.

Still have to sort out the window dimensions/resizing issue, and a few other bugs.

delroth commented:
@linkmauve would you mind giving this a test / review / help? :)

Sunderland93 commented:
wl_shell is deprecated. Use xdg-shell instead.

Sunderland93 commented:
Any progress?

Sunderland93 commented:
@stenzek  Any news? It would be pretty cool to have full Wayland support in Dolphin 6.0

stenzek commented:
Moved over to using xdg-shell, and the decoration manager protocol (this gets us borders/controls for NoGUI). Window size is no longer hardcoded, and resizing is supported.

The window decorations will still disappear when running with the qt frontend under gnome, but work fine on kwin. Given Qt doesn't use the wayland QPA when running under gnome anyway.. ¯\_(ツ)_/¯

orbea commented:
> I'm still not sure about generating the headers and including them in the tree (approx. 2500 of the lines are from these), we might be better off generating them as part of the build process. Does anyone know what the standard practice here is?

Not sure how to do this in cmake, but we use this script which is executed by our hand made configure script to do this in RetroArch.

https://github.com/libretro/RetroArch/blob/master/gfx/common/wayland/generate_wayland_protos.sh

It will require wayland-scanner and wayland-protocols to accomplish, the latter can be added to external. We fallback to our own copy if the user's version is not installed.

stenzek commented:
Yeah, I was considering using cmake to shell out to wayland-scanner.. getting the location on system without hardcoding it to /usr/share or whatever could be interesting.

orbea commented:
I was worried about that too so I made it check `$SHARE_DIR/wayland-protocols`, `/usr/local/share/wayland-protocols` and `/usr/share/wayland-protocols` in that order where `$SHARE_DIR` is the output of `--datarootdir`. I am not sure that is perfect, but so far I haven't seen anyone complain about it not working. However it will fallback to our `deps/wayland-protocols` if either none of them are found or our earlier check for version `>= 1.15` of `wayland-protocols.pc` fails (Using `pkg-config/pkgconf`). I've been told that earlier versions are not good enough and not all distros carry recent versions...

linkmauve commented:
> I'm still not sure about generating the headers and including them in the tree (approx. 2500 of the lines are from these), we might be better off generating them as part of the build process. Does anyone know what the standard practice here is?

extra-cmake-modules contains cmake scripts to do the latter, and is part of
every distribution worth supporting.

You can see how we use it in GLFW for instance, here we include ECM:
https://github.com/glfw/glfw/blob/master/CMakeLists.txt#L243

And here we use it to build protocol files:
https://github.com/glfw/glfw/blob/master/src/CMakeLists.txt#L34

It doesn’t do anything different from calling wayland-scanner manually,
but that still avoids having to handle e.g. finding the right binary to
use, or the protocol files.

stenzek commented:
@linkmauve thanks for the tip, I did come across the ECM modules but wasn't sure how widespread the support was. Sounds like the most straightforward method.

soreau commented:
[Here's what I've come up with so far.](https://github.com/soreau/dolphin-emu/tree/wayland-egl-2019) It uses subsurfaces and runs but

- Size is hard coded to 640x480. Need to get the inner dimensions of the render window somehow and the size when it goes fullscreen.
- Likewise, the offsets passed to wl_subsurface_set_position() are hard coded, so the emulation window isn't perfectly affixed to the render window providing the decorations.
- Moving the pointer over the emulation window causes a crash with the message `QObject::connect(QtWaylandClient::QWaylandWindow, QtWaylandClient::QWaylandInputDevice::Pointer): invalid null parameter` and I don't know why.
- Vulkan currently does not work, only GL does.

There's also the matter of hooking up keybindings etc. but since it runs, I'd say it's about half way there. Any help solving these problems is welcome.

stenzek commented:
I decided against using subsurfaces for this case, because to me it seems like it creates a bunch of problems, e.g. what size are the decorations? are the decorations round or square? I also wanted to avoid having a large chunk of wayland code in the qt frontend.

I'm guessing Qt creates subsurfaces itself for the render-to-main-window case, because this works fine in my branch when taking over the surface for Dolphin's rendering. The separate window works as expected with server-side decorations.

> Size is hard coded to 640x480. Need to get the inner dimensions of the render window somehow and the size when it goes fullscreen.

Implemented in this PR. The size is passed as part of the WSI info at startup, with resize events being passed to the backend, updating the surface as needed. See https://github.com/dolphin-emu/dolphin/pull/7452/commits/e81c726ed5d1104b3f2676510b13f2dbca43c677 and https://github.com/dolphin-emu/dolphin/pull/7452/commits/6c8aec029f4b040d385b9e3935cc43eb4e663ad0

> Vulkan currently does not work, only GL does.

Also implemented in this PR.

soreau commented:
Great. You might check the [cmake code](https://github.com/soreau/dolphin-emu/blob/wayland-egl-2019/CMakeLists.txt#L447) I have, it might give you some ideas.

soreau commented:
@stenzek I tested your code, it built fine and works (mostly). Mousing over and Vulkan with Qt is good but trying nogui ended up with xwayland. The keyboard did not work except in fullscreen, Esc caused crash and Alt+Enter got out of fullscreen but restarted emulation and left the main gui unresponsive (had to ctrl+c to stop emulation and exit). The keyboard should be hooked up for save/load state, Esc to quit properly and toggle fullscreen etc. Click to move emulation window would be nice to have too. It's a shame to see subsurfaces are more or less a waste of time but at least SSD is (apparently) an option now.

stenzek commented:
> Great. You might check the cmake code I have, it might give you some ideas.

I've been planning to move it over to the ECM module, just haven't had time yet.

> Mousing over and Vulkan with Qt is good but trying nogui ended up with xwayland.

`dolphin-emu-nogui -p wayland ...`. It'll default to X11 otherwise.

stenzek commented:
Now using ECM's scripts for wayland, as suggested by @linkmauve. I think this can be considered complete now, since CSD is outside the scope.

soreau commented:
It pulls in xkbcommon without hooking up the keyboard. Keyboard bindings need to be hooked up still. I'm not entirely sure but judging from latest simple-egl.c in weston, xkbcommon isn't needed to get key events.

soreau commented:
Another problem I have found is that after building (with wayland enabled), fullscreen no longer works in x11. With wayland disabled, emulation video is broken with x11. The window doesn't come up when audio starts and if it does show up, it only displays a single frame and never updates. So this PR breaks 'normal' x11 functionality quite badly.

stenzek commented:
It is hooked up to the keyboard. See https://github.com/dolphin-emu/dolphin/pull/7452/commits/6d51b712eb640b0310daefe477a38c61965ed753. Did you change the input device in the controller options? Wayland will show up as a different device to xinput.

I'll look into the X11 issues.

soreau commented:
Now that I changed the keyboard in hotkey config, the keybindings work. One other bug I noticed though, when I press Super key (while running under wayland) while in main dolphin window when emulation is not running, dolphin segfaults (regardless what 'keyboard' I select). I built with debug symbols to get this back trace http://ix.io/1HSQ

stenzek commented:
@soreau I'm unable to reproduce any of the X11 issues you've mentioned, regardless of egl/wayland/etc config options.

soreau commented:
@stenzek After trying various things I did manage to get it 'working' by disabling compositing window manager but it's extra slow (tww) and dolphin 5c5e6df0380 works fine with compositing at expected speeds. With your code, while the compositing wm is running and the emulation window displaying garbage, the audio is still slowed. Are you using a compositing window manager? I am surprised you can't reproduce because the problem is obvious on my end.

stenzek commented:
I tried both kwin and mutter? (or whatever the gnome compositor is) on Ubuntu 19.04.

Helios747 commented:
Worth noting that the earliest ubuntu that will support this without PPAs is 19.04. This shouldn't be a problem once CMake can just flip it on/off at configure time but for people not on rolling release distros, it may be worth sticking a small note in the README for ubuntu users hoping to leverage wayland support.

soreau commented:
@stenzek You sure you weren't using kwin and mutter wayland compositors? I'll try to bisect on my end if I can find a good commit.

soreau commented:
I got 0cb27cc42cfbfee78b16ea436ca60cd9d2bd6086 is the first bad commit. From the code "// Stop filling the background once emulation starts, but fill it until then (Bug 10958)". That comment doesn't make much sense but it certainly is what I am experiencing. The emulation window is never filled with the pixels. I suppose I could file an issue for it.

stenzek commented:
Interesting. Which qt version are you building against?

soreau commented:
This is with 4.8.7. Now that you said this, I tried 5.13.0 and results are better, the emulation starts but if I try to fullscreen, it stops rendering and even exiting fullscreen does not bring it back. The non-fullscreen window remains stuck on the screen and can't drag by titlebar.

soreau commented:
I've filed https://bugs.dolphin-emu.org/issues/11711

soreau commented:
I meant s/4.8.7/5.9.5/

soreau commented:
The x11 problem is apparently specific to my setup, I thought it probably had something to do with this PR. Now the only real problem I see left is the Qt frontend crashing when I press the Super key with the backtrace I posted previously but it seems to only happen in wayfire, not weston.

jordan-woyak commented on Source/Core/InputCommon/ControllerInterface/Wayland/Wayland.cpp:
```diff
@@ -0,0 +1,386 @@
+
+std::string KeyboardMouse::GetName() const
+{
+  return "Seat";
```
Despite perhaps being accurate, "Seat" is maybe a weird name to display to our users.

soreau commented:
Another crash I have found, though it only happens in wayfire, not in weston or rootston: If hotkey device is set to `[disconnected] XInput2/0/Virtual core pointer` and while under wayland I set it to `Wayland/0/Seat` and click Close, it crashes with this bt: http://ix.io/1Ihy After this crash, I can go to the hotkey window and click Close again and it does not crash. It only happens on the first time changing from the xinput to wayland seat. Probably something specific to wayfire, though I figured I'd mention it.

soreau commented:
@stenzek Here's a patch to fix the aforementioned crash: http://ix.io/1IhF

I think moving the window with keybinding causes key press events to be queued and when clicking Close, it actually sets up the keyboard_listener and gets those events but AddKeyInputs() hasn't been called. In the case of weston, AddKeyInputs() is called up front and also gets called again after closing the hotkey window and clicking on the main window. With wlroots compositors, AddKeyInputs() isn't called upfront and this leads to the problem. I'm still investigating why this is exactly.

stenzek commented:
Hm, interesting, the dispatch/roundtrip combo here should ensure that the keymap callback is fired before the inputs are added: https://github.com/dolphin-emu/dolphin/pull/7452/files#diff-f69d44ea5b6132331ebf06703f03fd9eR94

Perhaps this is an incorrect assumption..

soreau commented:
@stenzek Yes, with weston, the keymap is found by those calls but with wlroots compositors, the keymap is found in later event iterations.

soreau commented:
@stenzek Perhaps something like this would be the best solution: http://ix.io/1IhN

soreau commented:
@stenzek Take two: http://ix.io/1Ewq

After chatting with a wlroots dev, this is about the best we can get. This does dispatch+roundtrip+dispatch to get the keyboard listener setup and then if we have a listener, keep dispatching until we get the keymap (or we're disconnected).

stenzek commented:
Okay, thanks, will look into it when I have some time. Which compositor are you using?

soreau commented:
I've been testing with rootston, wayfire and weston.

stenzek commented:
@soreau patch is applied. I've also changed the input device name to "Keyboard and Mouse".

Tilka commented:
When forcing `QT_QPA_PLATFORM=wayland` I get warnings like these:
```
qt.qpa.wayland: Non-toplevel surfaces can't request window states
qt.qpa.wayland: Wayland does not support QWindow::requestActivate()
```

soreau commented:
@Tilka I don't get this error under weston or wayfire but it might be harmless either way. Which compositor are you using?

Tilka commented:
I'm using gnome-shell/mutter on up-to-date Arch Linux. Starting normally I get `Warning: Ignoring XDG_SESSION_TYPE=wayland on Gnome. Use QT_QPA_PLATFORM=wayland to run on Wayland anyway.` Doing that, I get the following issues in addition to the warnings mentioned above:

- The render window isn't recognized as being part of the same application, so Alt+key-above-tab doesn't work.
- Window decorations, standard dialogs, standard icons etc. don't match the GNOME theme.
- I have to change input devices in both the controller dialog and the hotkey dialog. Would it make sense to do this automatically?

soreau commented:
> I'm using gnome-shell/mutter on up-to-date Arch Linux.

Maybe this is why you get the qt messages.

>     * The render window isn't recognized as being part of the same application, so Alt+key-above-tab doesn't work.
I wouldn't worry much about window-management specific issues so much, they could be different depending on the compositor you choose.
>     * Window decorations, standard dialogs, standard icons etc. don't match the GNOME theme.
This is because such windows are using client side decorations. GTK and Qt each choose their own themes and there's not much that can be done unless each chooses to participate in a protocol that would allow the compositor to render unified decorations (with a single theme).
>     * I have to change input devices in both the controller dialog and the hotkey dialog. Would it make sense to do this automatically?
@stenzek ^^

Tilka commented:
Some more (probably also Qt or compositor issues):
- The application title in the tab switch popup is shown as "org.dolphin-em..." and there is no icon.
- The icon in the window's titlebar is pixelated and horizontally squished.

> This is because such windows are using client side decorations. GTK and Qt each choose their own themes and there's not much that can be done unless each chooses to participate in a protocol that would allow the compositor to render unified decorations (with a single theme).

I know, I'm just surprised this is still a problem, but I guess that's why it's not enabled by default.

soreau commented:
> Some more (probably also Qt or compositor issues):
> 
>     * The application title in the tab switch popup is shown as "org.dolphin-em..." and there is no icon.
> 
>     * The icon in the window's titlebar is pixelated and horizontally squished.

Yes, these are compositor/DE details.

> 
> > This is because such windows are using client side decorations. GTK and Qt each choose their own themes and there's not much that can be done unless each chooses to participate in a protocol that would allow the compositor to render unified decorations (with a single theme).
> 
> I know, I'm just surprised this is still a problem, but I guess that's why it's not enabled by default.

Be thankful if there are no rendering issues and the window is resizable.

Does render-in-main-window work for you?

Tilka commented:
> Does render-in-main-window work for you?

No, the render area turns transparent. Additionally, Dolphin crashes when switching to fullscreen in that mode (when rendering to a separate window it works):

```
#0  0x00007fffef89a755 in raise () from /usr/lib/libc.so.6
#1  0x00007fffef885851 in abort () from /usr/lib/libc.so.6
#2  0x00007ffff06d5fe6 in llvm::report_bad_alloc_error(char const*, bool) () from /usr/lib/libLLVM-8.so
#3  0x00007fffefc0bae2 in operator new (sz=140737213746017) at /build/gcc/src/gcc/libstdc++-v3/libsupc++/new_op.cc:55
#4  0x000055555569b33d in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*> (this=0x7fffbfffe038, __beg=0x555557a61150 "P", __end=<optimized out>)
    at /usr/include/c++/9.1.0/bits/char_traits.h:363
#5  0x0000555555c3ebaa in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct_aux<char const*> (__end=<optimized out>, __beg=<optimized out>, this=0x7fffbfffe038)
    at /usr/include/c++/9.1.0/bits/basic_string.h:263
#6  std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char const*> (__end=<optimized out>, __beg=<optimized out>, this=0x7fffbfffe038)
    at /usr/include/c++/9.1.0/bits/basic_string.h:266
#7  std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string (__a=..., __n=<optimized out>, __s=<optimized out>, this=0x7fffbfffe038)
    at /usr/include/c++/9.1.0/bits/basic_string.h:513
#8  std::__cxx11::basic_stringbuf<char, std::char_traits<char>, std::allocator<char> >::basic_stringbuf (__mode=std::_S_in, __str="", this=0x7fffbfffdff0) at /usr/include/c++/9.1.0/sstream:127
#9  std::__cxx11::basic_istringstream<char, std::char_traits<char>, std::allocator<char> >::basic_istringstream (__mode=std::_S_in, __str="", this=0x7fffbfffdfe0, __in_chrg=<optimized out>, 
    __vtt_parm=<optimized out>) at /usr/include/c++/9.1.0/sstream:462
#10 TryParse<float> (
    str="P\000\000\000\000\000\000\000a\000\000\000\000\000\000\000Pˡ\357\377\177\000\000\200\334\372WUU\000\000\030\000\000\000\000\000\000\000M\000+\000 \000\062\000\060\000\000\000\000\000\000\000\060\000\000\000\000\000\000\000p@\364WUU\000\000\006\000\000\000\000\000\000\000\030\000\000\000\000\000\000\000A\000r\000i\000a\000`\000\000\000\000\000\000\000 \000\000\000\000\000\000\000Mono\000U\000\000\000\000\000\000\000\000\000\000 \000\000\000\000\000\000\000\221\001\000\000\000\000\000\000\000\276\375WUU\000\000\020\062\374WUU\000\000\030\000\000\000\000\000\000\000M\000+\000 \000\062\000\060\000\000\000\000\000\000\000P\000\000\000\000\000\000\000\260\006\372WUU\000\000"..., output=output@entry=0x7fffbfffe1c4) at ../Source/Core/Common/StringUtil.h:58
#11 0x0000555555c3ed08 in TryParse (
    str="P\000\000\000\000\000\000\000a\000\000\000\000\000\000\000Pˡ\357\377\177\000\000\200\334\372WUU\000\000\030\000\000\000\000\000\000\000M\000+\000 \000\062\000\060\000\000\000\000\000\000\000\060\000\000\000\000\000\000\000p@\364WUU\000\000\006\000\000\000\000\000\000\000\030\000\000\000\000\000\000\000A\000r\000i\000a\000`\000\000\000\000\000\000\000 \000\000\000\000\000\000\000Mono\000U\000\000\000\000\000\000\000\000\000\000 \000\000\000\000\000\000\000\221\001\000\000\000\000\000\000\000\276\375WUU\000\000\020\062\374WUU\000\000\030\000\000\000\000\000\000\000M\000+\000 \000\062\000\060\000\000\000\000\000\000\000P\000\000\000\000\000\000\000\260\006\372WUU\000\000"..., output=output@entry=0x7fffbfffe200) at ../Source/Core/Common/StringUtil.cpp:286
#12 0x00005555556985cc in Config::detail::TryParse<bool, (void*)0> (
    str_value="P\000\000\000\000\000\000\000a\000\000\000\000\000\000\000Pˡ\357\377\177\000\000\200\334\372WUU\000\000\030\000\000\000\000\000\000\000M\000+\000 \000\062\000\060\000\000\000\000\000\000\000\060\000\000\000\000\000\000\000p@\364WUU\000\000\006\000\000\000\000\000\000\000\030\000\000\000\000\000\000\000A\000r\000i\000a\000`\000\000\000\000\000\000\000 \000\000\000\000\000\000\000Mono\000U\000\000\000\000\000\000\000\000\000\000 \000\000\000\000\000\000\000\221\001\000\000\000\000\000\000\000\276\375WUU\000\000\020\062\374WUU\000\000\030\000\000\000\000\000\000\000M\000+\000 \000\062\000\060\000\000\000\000\000\000\000P\000\000\000\000\000\000\000\260\006\372WUU\000\000"...) at ../Source/Core/Common/Config/Layer.h:23
#13 Config::Layer::Get<bool> (location=..., this=0x5555572c2f90) at ../Source/Core/Common/Config/Layer.h:117
#14 Config::Layer::Get<bool> (config_info=..., this=0x5555572c2f90) at ../Source/Core/Common/Config/Layer.h:108
#15 Config::Get<bool> (info=...) at ../Source/Core/Common/Config/Config.h:57
#16 0x000055555584f859 in VideoInterface::Update (ticks=2311024716) at ../Source/Core/Core/HW/VideoInterface.cpp:742
```

Tilka commented:
Also, running Dolphin on a fresh user config without ever opening the controller dialog:
```
#0  0x0000555555afc7f7 in ciface::Wayland::KeyboardMouse::KeyboardKey (data=0x7fffb88d5320, keyboard=0x7fffd0001760, serial=<optimized out>, time=<optimized out>, key=56, state=1)
    at /usr/include/c++/9.1.0/bits/stl_vector.h:1040
#1  0x00007fffea8696d0 in ffi_call_unix64 () from /usr/lib/libffi.so.6
#2  0x00007fffea8690a0 in ffi_call () from /usr/lib/libffi.so.6
#3  0x00007ffff4992f8f in ?? () from /usr/lib/libwayland-client.so.0
#4  0x00007ffff498f6ba in ?? () from /usr/lib/libwayland-client.so.0
#5  0x00007ffff4990bfc in wl_display_dispatch_queue_pending () from /usr/lib/libwayland-client.so.0
#6  0x0000555555afc9ad in ciface::Wayland::KeyboardMouse::UpdateInput (this=0x7fffb88d5320) at ../Source/Core/InputCommon/ControllerInterface/Wayland/Wayland.cpp:134
#7  0x0000555555af27e5 in ControllerInterface::UpdateInput (this=0x555556ae6160 <g_controller_interface>) at /usr/include/c++/9.1.0/bits/shared_ptr_base.h:1020
#8  0x0000555555848ddc in SerialInterface::UpdateDevices () at ../Source/Core/Core/HW/SI/SI.cpp:634
#9  0x000055555584f84d in VideoInterface::Update (ticks=498362436) at ../Source/Core/Core/HW/VideoInterface.cpp:738
```

soreau commented:
I had a similar experience trying to run-in-main-window here, the render area went transparent and dolphin unresponsive. Thanks for testing and the bt's, hopefully @stenzek will have a look.

Tilka commented:
On this fresh config fullscreen with render-to-main crashes in a different way:
```
qt.qpa.wayland.backingstore: Delivering update request through fallback timer, may not be in sync with display
wl_display@1: error 0: invalid object 58
The Wayland connection experienced a fatal error (Invalid argument)
[Thread 0x7fffc7fff700 (LWP 4563) exited]
[Thread 0x7fffd6ffd700 (LWP 4560) exited]
[Thread 0x7fffe5748700 (LWP 4537) exited]
terminate called without an active exception

Thread 1 "dolphin-emu" received signal SIGABRT, Aborted.
0x00007fffef89a755 in raise () from /usr/lib/libc.so.6
=> 0x00007fffef89a755 <raise+325>:	48 8b 84 24 08 01 00 00	mov    0x108(%rsp),%rax
(gdb) bt
#0  0x00007fffef89a755 in raise () from /usr/lib/libc.so.6
#1  0x00007fffef885851 in abort () from /usr/lib/libc.so.6
#2  0x00007fffefbfe81f in __gnu_cxx::__verbose_terminate_handler () at /build/gcc/src/gcc/libstdc++-v3/libsupc++/vterminate.cc:95
#3  0x00007fffefc0b30a in __cxxabiv1::__terminate (handler=<optimized out>) at /build/gcc/src/gcc/libstdc++-v3/libsupc++/eh_terminate.cc:47
#4  0x00007fffefc0b367 in std::terminate () at /build/gcc/src/gcc/libstdc++-v3/libsupc++/eh_terminate.cc:57
#5  0x00005555557afa8d in std::thread::~thread (this=<optimized out>, __in_chrg=<optimized out>) at /usr/include/c++/9.1.0/thread:139
#6  0x00007fffef89ce70 in __run_exit_handlers () from /usr/lib/libc.so.6
#7  0x00007fffef89cfae in exit () from /usr/lib/libc.so.6
#8  0x00007fffe65f4ebf in QtWaylandClient::QWaylandDisplay::exitWithError() () from /usr/lib/libQt5WaylandClient.so.5
#9  0x00007fffe65f4f1a in QtWaylandClient::QWaylandDisplay::flushRequests() () from /usr/lib/libQt5WaylandClient.so.5
#10 0x00007ffff4dc7a11 in QMetaObject::activate(QObject*, int, int, void**) () from /usr/lib/libQt5Core.so.5
#11 0x00007ffff4dd45f2 in QSocketNotifier::activated(int, QSocketNotifier::QPrivateSignal) () from /usr/lib/libQt5Core.so.5
#12 0x00007ffff4dd4942 in QSocketNotifier::event(QEvent*) () from /usr/lib/libQt5Core.so.5
#13 0x00007ffff7a3e4d5 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /usr/lib/libQt5Widgets.so.5
#14 0x00007ffff7a47a21 in QApplication::notify(QObject*, QEvent*) () from /usr/lib/libQt5Widgets.so.5
#15 0x00007ffff4d9b9c2 in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /usr/lib/libQt5Core.so.5
#16 0x00007ffff4df45f6 in ?? () from /usr/lib/libQt5Core.so.5
#17 0x00007fffeae42cf4 in g_main_context_dispatch () from /usr/lib/libglib-2.0.so.0
#18 0x00007fffeae44b11 in ?? () from /usr/lib/libglib-2.0.so.0
#19 0x00007fffeae44b51 in g_main_context_iteration () from /usr/lib/libglib-2.0.so.0
#20 0x00007ffff4df39a3 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/libQt5Core.so.5
#21 0x00007ffff4d9a5ec in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/libQt5Core.so.5
#22 0x00007ffff4da2326 in QCoreApplication::exec() () from /usr/lib/libQt5Core.so.5
#23 0x000055555565c159 in main (argc=<optimized out>, argv=<optimized out>) at ../Source/Core/DolphinQt/Main.cpp:223
```

jackoalan commented:
@stenzek what's the status of your wayland branch? I know there are some merge conflicts right now, is it ready to be worked on? I'd gladly make the requested changes:
https://github.com/dolphin-emu/dolphin/pull/8727#issuecomment-610285750

It also looks like your changes in f081713 are workarounds to avoid widget hierarchy changes. These would no longer be necessary with proper surface destroy/create handling.

stenzek commented:
Rebased. No idea if it'll still work, when I have time I'll check.

jackoalan commented:
Cool! I'll give it a shot

jackoalan commented:
Also, do you think pausing the emulator is really necessary as opposed to using condition variables? The whole surface change cycle doesn't take more than 100ms on KWin, and it's not like Dolphin's UI Thread is being hammered super hard to cause a delay between two events. Such delays are bound to occur on other platforms' surface creators anyhow.

newperson1746 commented:
Any chance of this making it into master? Fedora has had Wayland default for years and many other distros using gnome-shell are eyeing a Wayland default for their next releases

LaserEyess commented:
@stenzek could this be rebased? It does not build against gcc 10.0.

Rumi-Larry commented:
Pokey, poke.

newperson1746 commented:
Oh hi Stenzek, good to see you back on this PR! Are there any tweaks left we could help with?

LaserEyess commented:
I got a crash when pressing ctrl+q to quit. Unfortunately the trace is not very helpful since I didn't have libwayland debug symbols. I will try to reproduce.

https://0x0.st/-AlV.txt

soreau commented:
I notice this when building:

[ 80%] Generating wayland-xdg-shell-protocol.c
[ 80%] Generating wayland-xdg-decoration-protocol.c
Using "code" is deprecated - use private-code or public-code.
See the help page for details.
Using "code" is deprecated - use private-code or public-code.
See the help page for details.

[Reference](https://gitlab.freedesktop.org/wayland/wayland/-/blob/master/src/scanner.c#L76-78)

soreau commented:
> I got a crash when pressing ctrl+q to quit. Unfortunately the trace is not very helpful since I didn't have libwayland debug symbols. I will try to reproduce.
> 
> https://0x0.st/-AlV.txt

[Here's what I get](http://ix.io/2WLE) with Ctrl+Q. I didn't see this in the Dolphin keybinding settings which makes me wonder if Qt is listening and taking action, and Dolphin isn't expecting it.

soreau commented:
Also in my case, even if all the required wayland development packages are installed, wayland is still not automatically enabled. I had to manually enable wayland with `-DENABLE_WAYLAND=1` or else there is this message when attempting to start emulation:

![snapshot](https://user-images.githubusercontent.com/1450125/115471129-b83c8c00-a1f4-11eb-9018-ff36b4d96539.png)

LaserEyess commented:
ctrl+q is a shortcut to send sigterm to the focused window on my compositor, and a general shortcut on many qt/qtk programs. If dolphin itself doesn't implement it as a shortcut it should still handle it properly.

v1993 commented:
I think it makes sense to enable Wayland by default until there's a good reason not to.

tobiasjakobi commented:
Just voicing my interest in this PR as a fellow Wayland (sway) user.

devomaa commented:
+1 any progress on this?

MatthewCroughan commented:
If anybody wants to build or run this PR using [`nix`](https://nixos.org/download.html), they can do so like this:
### Build (Puts it in ./result/bin/dolphin-emu)
`nix build github:matthewcroughan/nixpkgs/dolphin-wayland-hack#dolphinEmuMaster`
### Run (Just runs `dolphin-emu`)
`nix shell github:matthewcroughan/nixpkgs/dolphin-wayland-hack#dolphinEmuMaster --command dolphin-emu`

Need4Speed402 commented:
I created a .patch that is based on this pr's work and rebased on 6e7698a3540eb3e9186bea442f08d5d96544d73f. Here's the patch:
https://pastebin.com/tbGEQY31.

My efforts were like a monkey smashing a keyboard. I just fixed some function signatures and it seems to compile on gcc11. I'm not sure how stable it is, but it seems to launch opengl/vulkan on wayland now on a couple of games I tried. This is running archlinux and sway (HiDPI seems to work as well)
I did make one change from the original patch, and that is to have cmake enable wayland by default. Before you had to set the ENABLE_WAYLAND option. I just decided to do that because people using this .patch would probably already want to be doing that.

MatthewCroughan commented:
@Need4Speed402 Thank you so much. I have made a commit here https://github.com/MatthewCroughan/nixpkgs/commit/3863ef16509da9288d85d5c3d6320329dad7e54d that will allow anyone viewing this PR to use Nix on any distribution (including Arch :smile:) to test your patch. I've tested it as working under Sway on NixOS too.

### Build Need4Speed402's patch (Puts it in ./result/bin/dolphin-emu)
`nix build github:matthewcroughan/nixpkgs/dolphin-wayland-hack-n4s#dolphinEmuMaster`
### Run Need4Speed402's patch (Just runs `dolphin-emu` after building it)
`nix shell github:matthewcroughan/nixpkgs/dolphin-wayland-hack-n4s#dolphinEmuMaster --command dolphin-emu`

4e554c4c commented:
Any indication on why this was closed?
Is the patch nonviable or does it just need someone to clean it up before it can be pulled?

AdmiralCurtiss commented:
Feel free to rebase this as a new PR if you're willing to get this into a mergeable state.

soreau commented:
> Feel free to rebase this as a new PR if you're willing to get this into a mergeable state.

It would be nice to know from upstream maintainers if this work would even be accepted if such a PR exists, before doing the work.

JosJuice commented:
> Any indication on why this was closed?

Stenzek closed all his open PRs. So it's not because something was wrong with this PR specifically.

Need4Speed402 commented:
I did rebase this PR recently [here](https://github.com/dolphin-emu/dolphin/pull/7452#issuecomment-882307772) and it seems to apply cleanly on current dolphin still. I have played some games through it and I can say that the only issue is really just the mouse cursor not hiding itself when it should be. I know there was some chit chat  about the PR not rendering window decorations on Gnome? Maybe somebody should test that. I would be willing to clean this up and put up a new PR but life's kinda busy right now so I'd rather not.

iMonZ commented:
Any news here?

Nefsen402 commented:
I rebased the [patches](https://github.com/dolphin-emu/dolphin/files/7836816/wayland_support.txt) again on commit 3da64875314e2a1743b8d6453403ef85e8b45eb4.  Works fine on sway git.

iMonZ commented:
> I rebased the [patches](https://github.com/dolphin-emu/dolphin/files/7836816/wayland_support.txt) again on commit [3da6487](https://github.com/dolphin-emu/dolphin/commit/3da64875314e2a1743b8d6453403ef85e8b45eb4). Works fine on sway git.

Thanks.
So what step should be done afterwards?

TheSunCat commented:
What's holding this back from being merged currently?

tobiasjakobi commented:
> What's holding this back from being merged currently?

The fact that this PR is closed.
